runtime/contract_update_validation.go

```diff
@@ -0,0 +1,456 @@
+	// This also avoids getting stuck on circular dependencies between composite decls.
+	if validator.visited[newDeclaration] {
+		return nil
+	}
+	validator.visited[newDeclaration] = true
+
+	err := validator.checkFields(oldDeclaration, newDeclaration)
+	if err != nil {
+		return err
+	}
+
+	err = validator.checkNestedDeclarations(oldDeclaration, newDeclaration)
+	if err != nil {
+		return err
+	}
+
+	switch newDecl := newDeclaration.(type) {
+	case *ast.CompositeDeclaration:
+		oldDecl := oldDeclaration.(*ast.CompositeDeclaration)
+		return validator.checkConformances(oldDecl, newDecl)
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) checkFields(
+	oldDeclaration ast.Declaration,
+	newDeclaration ast.Declaration,
+) error {
+
+	oldFields := oldDeclaration.DeclarationMembers().FieldsByIdentifier()
+	newFields := newDeclaration.DeclarationMembers().Fields()
+
+	// Updated contract has to have at-most the same number of field as the old contract.
+	// Any additional field may cause crashes/garbage-values when deserializing the
+	// already-stored data. However, having less number of fields is fine for now. It will
+	// only leave some unused-data, and will not do any harm to the programs that are running.
+
+	// This is a fail-fast check.
+	if len(oldFields) < len(newFields) {
+		currentDeclName := newDeclaration.DeclarationIdentifier()
+		return &TooManyFieldsError{
+			declName:       currentDeclName.Identifier,
+			expectedFields: len(oldFields),
+			foundFields:    len(newFields),
+			Range:          ast.NewRangeFromPositioned(currentDeclName),
+		}
+	}
+
+	for _, newField := range newFields {
+		oldField := oldFields[newField.Identifier.Identifier]
+		if oldField == nil {
+			return &ExtraneousFieldError{
+				declName:  newDeclaration.DeclarationIdentifier().Identifier,
+				fieldName: newField.Identifier.Identifier,
+				Range:     ast.NewRangeFromPositioned(newField.Identifier),
+			}
+		}
+
+		err := validator.checkField(newField, oldField)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) checkField(
+	newField *ast.FieldDeclaration,
+	oldField *ast.FieldDeclaration,
+) error {
+
+	oldType := oldField.TypeAnnotation.Type
+
+	err := oldType.CheckEqual(newField.TypeAnnotation.Type, validator)
+	if err != nil {
+		return &FieldMismatchError{
+			declName:  validator.currentDecl.DeclarationIdentifier().Identifier,
+			fieldName: newField.Identifier.Identifier,
+			err:       err,
+			Range:     ast.NewRangeFromPositioned(newField),
+		}
+	}
+	return nil
+}
+
+func (validator *ContractUpdateValidator) checkNestedDeclarations(
+	oldDeclaration ast.Declaration,
+	newDeclaration ast.Declaration,
+) error {
+
+	oldNestedCompositeDecls := oldDeclaration.DeclarationMembers().CompositesByIdentifier()
+	oldNestedInterfaceDecls := oldDeclaration.DeclarationMembers().InterfacesByIdentifier()
+
+	getCompositeOrInterfaceDecl := func(name string) (ast.Declaration, bool) {
```

Comment:
Maybe indicate that this function looks up the composite or interface declaration in the old nested declarations, e.g. rename to `getOldCompositeOrInterfaceDecl`